gschem/src/i_basic.c

```diff
@@ -207,6 +207,34 @@ void i_action_update_status (GschemToplevel *w_current, gboolean inside_action)
 void i_set_state(GschemToplevel *w_current, enum x_states newstate)
 {
   i_set_state_msg(w_current, newstate, NULL);
+  const gchar* mode;
+    switch (newstate) {
+      case SELECT: mode="select-mode"; break;
+      case GRIPS: mode="grips-mode"; break;
+      case ARCMODE: mode="arc-mode"; break;
+      case BOXMODE: mode="box-mode"; break;
+      case BUSMODE: mode="bus-mode"; break;
+      case CIRCLEMODE: mode="circle-mode"; break;
+      case LINEMODE: mode="line-mode"; break;
+      case NETMODE: mode="net-mode"; break;
+      case PATHMODE: mode="path-mode"; break;
+      case PICTUREMODE: mode="picture-mode"; break;
+      case PINMODE: mode="pin-mode"; break;
+      case COMPMODE: mode="comp-mode"; break;
+      case COPYMODE: mode="copy-mode"; break;
+      case MCOPYMODE: mode="multiple-copy-mode"; break;
+      case MOVEMODE: mode="move-mode"; break;
+      case PASTEMODE: mode="paste-mode"; break;
+      case TEXTMODE: mode="text-mode"; break;
+      case SBOX: mode="box-select-mode"; break;
+      case ZOOMBOX: mode="zoom-box-mode"; break;
+      case PAN: mode="pan-mode"; break;
+      case MIRRORMODE: mode="mirror-mode"; break;
+      case ROTATEMODE: mode="rotate-mode"; break;
+      default: mode="select-mode"; break;
```

Comment:
I suggest using a [`g_return_if_reached()`](https://people.gnome.org/~desrt/glib-docs/glib-Warnings-and-Assertions.html#g-return-if-reached) in the default case here, or have no default so that the compiler warns if the `switch` statement does not cover every possible value of `enum x_states`.  If you choose to have no `default:` label, I suggest initialising the `mode` variable to the appropriate default value, e.g.:

```c
const gchar* mode = "select-mode";
switch (newstate) {
    /* ... */
}
```

You don't need a `;` after the closing `}` of a `switch` statement.